[SPARK-14684] [SPARK-15277] [SQL] Partition Spec Validation in SessionCatalog and Checking Partition Spec Existence Before Dropping

gatorsmile opened the pull request:
#### What changes were proposed in this pull request?

~~Currently, multiple partitions are allowed to drop by using a single DDL command: Alter Table Drop Partition. However, the internal implementation could break atomicity. That means, we could just drop a subset of qualified partitions, if hitting an exception when dropping one of qualified partitions~~

~~This PR contains the following behavior changes:~~
~~\- disallow dropping multiple partitions by a single command ~~
~~- allow users to input predicates in partition specification and issue a nicer error message if the predicate's comparison operator is not `=`.~~
~~\- verify the partition spec in SessionCatalog. This can ensure each partition spec in `Drop Partition` does not correspond to multiple partitions.~~

This PR has two major parts:
- Verify the partition spec in SessionCatalog for fixing the following issue:
  
  ``` scala
  sql(s"ALTER TABLE $externalTab DROP PARTITION (ds='2008-04-09', unknownCol='12')")
  ```
  
  Above example uses an invalid partition spec. Without this PR, we will drop all the partitions. The reason is Hive megastores getPartitions API returns all the partitions if we provide an invalid spec.
- Re-implemented the `dropPartitions` in `HiveClientImpl`. Now, we always check if all the user-specified partition specs exist before attempting to drop the partitions. Previously, we start drop the partition before completing checking the existence of all the partition specs. If any failure happened after we start to drop the partitions, we will log an error message to indicate which partitions have been dropped and which partitions have not been dropped. 
#### How was this patch tested?

Modified the existing test cases and added new test cases.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
SPARK-11633

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into joinMakeCopy

gatorspark committed:
Merge remote-tracking branch 'upstream/master'

gatorspark committed:
Merge branch 'master' of https://github.com/gatorsmile/spark

Commit:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
converge

gatorsmile committed:
converge

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
initial fix.

gatorsmile committed:
initial fix.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into banDropMultiPart

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into banDropMultiPart

gatorsmile committed:
fixed test cases

SparkQA commented:
**[Test build #57405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57405/consoleFull)** for PR 12801 at commit [`6272398`](https://github.com/apache/spark/commit/6272398e7ac97039a21b79a7361cca0d8250261f).

SparkQA commented:
**[Test build #57405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57405/consoleFull)** for PR 12801 at commit [`6272398`](https://github.com/apache/spark/commit/6272398e7ac97039a21b79a7361cca0d8250261f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57405/
Test PASSed.

gatorsmile commented:
cc @yhuai @andrewor14 @rxin

rxin commented:
The title doesn't match the pr description? Are you implementing this or disallowing it?

gatorsmile commented:
This PR is to disallow it. Sorry, update the PR description.

rxin commented:
I think the in-memory one and the Hive one should use the same behavior, and this should be enforced at the command level, not up to the specific implementation of the catalog.

gatorsmile commented:
I see. Will do it. Thanks!

gatorsmile committed:
address comments

SparkQA commented:
**[Test build #57458 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57458/consoleFull)** for PR 12801 at commit [`3b7b5de`](https://github.com/apache/spark/commit/3b7b5de9f48f02de514dc56f6995d0d6eeb16d24).

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge branch 'banDropMultiPart' into banDropMultiPartNew

# Conflicts:
#	sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala

SparkQA commented:
**[Test build #57460 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57460/consoleFull)** for PR 12801 at commit [`78a868b`](https://github.com/apache/spark/commit/78a868b1c6eca4351a61130ebc2796189c454a60).

SparkQA commented:
**[Test build #57458 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57458/consoleFull)** for PR 12801 at commit [`3b7b5de`](https://github.com/apache/spark/commit/3b7b5de9f48f02de514dc56f6995d0d6eeb16d24).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57458/
Test PASSed.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
**[Test build #57460 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57460/consoleFull)** for PR 12801 at commit [`78a868b`](https://github.com/apache/spark/commit/78a868b1c6eca4351a61130ebc2796189c454a60).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57460/
Test FAILed.

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:
```diff
@@ -81,8 +84,8 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
       tableType = CatalogTableType.EXTERNAL,
       storage = storage,
-      schema = Seq(),
-      createTime = 0L), ignoreIfExists = false)
```
... I need to revert it back

gatorsmile committed:
fix tests

SparkQA commented:
**[Test build #57481 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57481/consoleFull)** for PR 12801 at commit [`fa15228`](https://github.com/apache/spark/commit/fa152285763f9305e119f8cdce6116a11469cfe3).

SparkQA commented:
**[Test build #57481 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57481/consoleFull)** for PR 12801 at commit [`fa15228`](https://github.com/apache/spark/commit/fa152285763f9305e119f8cdce6116a11469cfe3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57481/
Test PASSed.

gatorsmile commented:
cc @rxin It is ready for review. : )

gatorsmile commented:
@cloud-fan Your PR https://github.com/apache/spark/pull/12871 is related to this. This is for disallowing users drop multiple partitions using one command. Could you please review this? Thanks!

andrewor14 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -248,6 +248,26 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
+      ctx: PartitionRangeSpecContext): Map[String, String] = withOrigin(ctx) {
+    ctx.partitionRangeVal.asScala.map { pVal =>
+      val operator = pVal.comparisonOperator().getChild(0).asInstanceOf[TerminalNode]
+      if (operator.getSymbol.getType != SqlBaseParser.EQ) {
```
I don't think this is worth doing in this patch if we just parse it to throw an exception. If we decide to support ranged partition specs later we can add it then, but now this just creates unnecessary code. I would all the parser changes in this patch.

gatorsmile replied:
I will revert it. The purpose here is to issue a better exception message if users specify predicates in alter table drop partition.

andrewor14 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -448,11 +454,19 @@ class SessionCatalog(
+    val tableMetadata = getTableMetadata(tableName)
+    // Verify if each part exactly matches the partition spec of the table definition
+    // Here, we disallow users to drop multiple partition using one spec for atomicity.
+    // For example, the partial spec a=1 could contains multiple partitions a=1/b=1 and a=1/b=2
```
this comment is really confusing. I thought the whole point of the patch is to disallow partial specs, e.g. `a=1` in this example?

gatorsmile replied:
I will remove all the comments.

andrewor14 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -510,6 +538,21 @@ class SessionCatalog(
+   * Verify if the input partition spec exactly matches the existing defined partition spec
+   * The columns must be the same but the orders could be different.
+   */
+  private def requireExactMatchedPartitionSpec(
```
I think the interface here could be better, maybe something like
```
private def requirePartitionsDefinedInTable(
      specs: Seq[TablePartitionSpec],
      table: CatalogTable): Unit = {
  ...
}
```
This would allow you to reuse more code

gatorsmile replied:
Sure, will do it.

gatorsmile replied:
The latest code changes have contained this great idea. Thanks!

andrewor14 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -465,6 +479,11 @@ class SessionCatalog(
+    // Verify if old and new specs exactly match the partition spec of the table definition
+    Seq(specs, newSpecs).flatten.foreach { s =>
+      requireExactMatchedPartitionSpec(s.keys.toSeq, tableMetadata.partitionColumnNames, tableName)
+    }
```
if you take my suggestion then this would become:
```
requirePartitionsDefinedInTables(specs, tableMetadata)
requirePartitionsDefinedInTables(newSpecs, tableMetadata)
```
a little simpler I think

gatorsmile replied:
Great! Thanks!

andrewor14 commented:
@gatorsmile I'm not sure if I completely understand the intent of this patch. If I understand correctly there are two ways to drop multiple partitions. If I have a table with partitions `a=1,b=1` and `a=1,c=2`, then:
(1) I can specify `a=1` and drop both partitions
(2) I can specify `a=1,b=1` and `a=1,c=2` explicitly and drop both partitions

It seems that your patch only guards against (1), right? There's still an "atomicity" problem here if I do (2). If that is the case then maybe this issue should be scoped differently.

gatorsmile commented:
@andrewor14 What is the partition spec if one table has two partitions `a=1`, `b=1` and `a=1`, `c=2`? I think it is impossible to do that, right? 

We enforce the partition spec of a drop partition command must strictly follow the partition spec of the table definition by the function `requireExactMatchedPartitionSpec ` https://github.com/apache/spark/pull/12801/files#diff-b3f9800839b9b9a1df9da9cbfc01adf8R545

Let me know if you have any question about this. Thanks!

cloud-fan commented:
is it allowed to do `ALTER TABLE tbl DROP PARTITION (a=1, b=1) PARTITION (a=1,c=2)`?

cloud-fan commented:
oh, looks like our parser forbid it. cc @andrewor14 , seems your second case is impossible?

gatorsmile commented:
@cloud-fan The parser still can do this if the partition spec of the table is `a, b, c`. However, the partial or invalid partition spec can be detected in this PR.

cloud-fan commented:
what if the partition spec of the table is a, b, and we do `ALTER TABLE tbl DROP PARTITION (a=1, b=1) PARTITION (a=1,b=2)`?

gatorsmile commented:
```scala
case class AlterTableDropPartition(
    tableName: TableIdentifier,
    specs: Seq[TablePartitionSpec],
    ifExists: Boolean)
```
The `specs` is a sequence of `TablePartitionSpec`. I think it is OK to do it.

gatorsmile commented:
But, this PR does not allow it for ensuring the atomicity of the command.
https://github.com/gatorsmile/spark/blob/banDropMultiPart/sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala#L425-L428

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge branch 'banDropMultiPartNew' into banDropMultiPartNewNew

# Conflicts:
#	sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4
#	sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLCommandSuite.scala
#	sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala

gatorsmile committed:
revert.

SparkQA commented:
**[Test build #57770 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57770/consoleFull)** for PR 12801 at commit [`8a4980c`](https://github.com/apache/spark/commit/8a4980c211c5ba95fa39719689632eafaec2c1f5).

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
**[Test build #57770 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57770/consoleFull)** for PR 12801 at commit [`8a4980c`](https://github.com/apache/spark/commit/8a4980c211c5ba95fa39719689632eafaec2c1f5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57770/
Test PASSed.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

andrewor14 commented:
What do you mean, I was able to drop multiple partitions in one command:
```
scala> sql("SHOW PARTITIONS my_tab1").show()
+-----------+
|     result|
+-----------+
|a=3/b=3/c=3|
|a=1/b=2/c=2|
|a=1/b=1/c=1|
+-----------+

scala> sql("ALTER TABLE my_tab1 DROP PARTITION (a=1, b=1, c=1), PARTITION (a=1, b=2, c=2)")

scala> sql("SHOW PARTITIONS my_tab1").show()
+-----------+
|     result|
+-----------+
|a=3/b=3/c=3|
+-----------+
```

andrewor14 commented:
What I'm saying is, the issue here is more like "removing support for partial specs when dropping partitions", NOT "disallow dropping multiple partitions in a single command". Right? The issue is not correctly scoped.

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -400,6 +400,10 @@ case class AlterTableDropPartition(
     }
+    if (specs.size > 1) {
+      throw new AnalysisException(
+        s"Alter Table Drop Partition is not allowed to drop more than one partitions.")
```
I see, now we throw an exception here to disallow multiple partitions. So why does this class even take in a sequence of specs int he first place? It should just take in a single spec. We should throw the exception in the parser so the code downstream can be cleaner.

andrewor14 replied:
actually I don't think we should throw an exception here. Let's move the discussion to the main thread.

andrewor14 commented:
Actually let's ping @yhuai. I personally don't think SPARK-15026 is a real problem. The user should still be able to drop multiple partitions at once. As for SPARK-14684, what is the bug in Hive that you're talking about? @gatorsmile Can you give a small reproduction for that?

gatorsmile commented:
@andrewor14 For the defect mentioned in SPARK-14684, you can reproduce it by using the following statement:
```
sql(s"ALTER TABLE $externalTab DROP PARTITION (ds='2008-04-09', unknownCol='12')")
```
Above example uses an invalid partition spec. Without this PR, we will drop all the partitions. The reason is Hive megastores `getPartitions` API returns all the partitions if we provide an invalid spec.

gatorsmile commented:
Regarding SPARK-15026, my major concern is we might get JIRAs if the previous `Drop Partition` command only drops a subset of partitions. If users issue the same command, they will get an error message:
```
org.apache.spark.sql.AnalysisException: partition to drop 'Map(ds -> 2008-04-08)' does not exist in table 'exttable_with_partitions' database 'default';
```

If you do not think this is a big deal, we need to issue a better error message. Let users know which partitions have been dropped and which partitions have not been dropped. This error message could be pretty long if one partial partition spec corresponds a long list of partitions. I am not sure if there exists any length limit of an exception message.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

yhuai commented:
@gatorsmile I am not sure we should ban dropping multiple partitions in a single call, which is a useful command. 

I just took a look at our implementation of addPartitions and dropPartitions in HiveClientImpl. Looks like we should use the batch version instead of asking drop every partition individually.

gatorsmile commented:
@yhuai Agree! Actually, I initially tried to do it in a batch mode. That is the following API we need to use
```
public List<Partition> dropPartitions(String dbName, String tblName, List<PartSpec> partSpecs, boolean deleteData, boolean ignoreProtection, boolean ifExists)
```
I hit an issue to generate `ExprNodeGenericFuncDesc` in `PartSpec`. 

The other `dropPartitions` APIs in `Hive.java` does not allow the caller to specify a partial spec. Any better idea? Thanks!

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

andrewor14 commented:
For now, let's just drop them one by one. Fixing that is a separate issue that we can always do later (even after 2.0)

gatorsmile commented:
I see. I will change the implementation of this PR. This PR will only fix SPARK-14684. 

Agree, we can address the atomicity issue later.

gatorsmile commented:
Will try to improve the existing error message if a multi-partition drop command failed. This can improve the usability before we supporting atomicity. Thanks!

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge remote-tracking branch 'upstream/master'

gatorsmile committed:
Merge branch 'banDropMultiPartNewNew' into banDropMultiPartNewNewNew

# Conflicts:
#	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala
#	sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala

gatorsmile committed:
address comments.

gatorsmile committed:
address comments.

gatorsmile committed:
issue a better error message

SparkQA commented:
**[Test build #58405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58405/consoleFull)** for PR 12801 at commit [`15f287f`](https://github.com/apache/spark/commit/15f287fa0101c091d7d81288e8abd75c19a29a9e).

gatorsmile commented:
**Update:** 
- Reverted the code changes.
    1. Users are allowed to specify the partial partition spec. For example, given a table with four partitions [a='1', b='x'], [a='2', b='x'], [a='1', b='y'], [a='2', b='y'], users can drop partitions by [a='1'] or [b='x]
    2. Users are also allowed to specify multiple partition specs in the same command. 
- Re-implemented the `dropPartitions` in `HiveClientImpl`. Now, we always check if all the user-specified partition specs exist before attempting to drop the partitions. Previously, we start drop the partition before completing checking the existence of all the partition specs.
- If any failure happened after we start to drop the partitions, we will log an error message to indicate which partitions have been dropped and which partitions have not been dropped. 
- Updated the checking logics of partition spec.

gatorsmile commented:
Just updated the title and PR description. The JIRA Spark-15026 is not valid now. Thus, created and included a new JIRA Spark-15277.

SparkQA commented:
**[Test build #58405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58405/consoleFull)** for PR 12801 at commit [`15f287f`](https://github.com/apache/spark/commit/15f287fa0101c091d7d81288e8abd75c19a29a9e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58405/
Test PASSed.

gatorsmile committed:
remove useless checks

SparkQA commented:
**[Test build #58423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58423/consoleFull)** for PR 12801 at commit [`9ca7621`](https://github.com/apache/spark/commit/9ca76217defe5e0154b1d26a61baf4831d19df3a).

SparkQA commented:
**[Test build #58423 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58423/consoleFull)** for PR 12801 at commit [`9ca7621`](https://github.com/apache/spark/commit/9ca76217defe5e0154b1d26a61baf4831d19df3a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58423/
Test PASSed.

gatorsmile commented:
@andrewor14 @cloud-fan @yhuai This is ready for review. Thanks!

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala:
```diff
@@ -565,6 +592,28 @@ class SessionCatalogSuite extends SparkFunSuite {
       ignoreIfNotExists = true)
   }
 
+  test("drop partitions with invalid partition spec") {
```
should we test drop partitions with `partWithLessColumns`?

gatorsmile replied:
Since In-Memory Catalog does not support it, the test will fail. Thus, I did not add it.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -595,6 +602,42 @@ class SessionCatalog(
 
+  /**
+   * Verify if the input partition spec exactly matches the existing defined partition spec
+   * The columns must be the same but the orders could be different.
```
Is it consistent with hive? Sorry I don't remember it clearly, but sometimes the partition order matters in hive?

gatorsmile replied:
Yeah, Hive supports it. To the external users, the orders do not matter. This is just for usability, I think.

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -405,20 +406,43 @@ private[hive] class HiveClientImpl(
-        dropOptions.ifExists = ignoreIfNotExists
-        client.dropPartition(db, table, hivePartition.getValues, dropOptions)
+    // do the check at first and collect all the matching partitions
+    val matchingParts =
```
should we support this feature in `InMemoryCatalog`?

gatorsmile replied:
So far, `InMemoryCatalog` does not correctly support partition dropping with a partial spec.

cloud-fan commented:
looks pretty good, thanks for working on it!

andrewor14 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:
```diff
@@ -523,13 +524,14 @@ class SessionCatalog(
-      parts: Seq[TablePartitionSpec],
+      specs: Seq[TablePartitionSpec],
       ignoreIfNotExists: Boolean): Unit = {
+    requirePartialMatchedPartitionSpec(specs, getTableMetadata(tableName))
```
Wait, I thought the intention of this patch is to ban the partial matching, which is buggy in Hive? Looks like this just ensures that the user doesn't specify a column that is not a partitioned column here. Is that sufficient to bypass the bug in Hive that you were talking about?

gatorsmile replied:
Yeah, if we do not specify the invalid partition spec, Hive can return a correct set of qualified partitions.

gatorsmile replied:
Thus, this PR resolves the issue by banning users to specify the invalid partition spec. 

Below is the test case:
https://github.com/gatorsmile/spark/blob/9ca76217defe5e0154b1d26a61baf4831d19df3a/sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala#L189-L194

Without this PR fix, it will drop all the partitions.

andrewor14 replied:
I understand now. The earlier PR descriptions were kind of confusing.

andrewor14 commented:
The latest changes look good so I'm merging them into master 2.0.

@gatorsmile In the future it would be good if you could summarize the issue with an example more succinctly from the beginning. This patch has gone through a few iterations and it wasn't clear to me what the Hive bug you were talking about is until later. The current PR description is much clearer.

gatorsmile commented:
Sure, will do. Thank you!
